idx/memory/priority_lock.go

```diff
@@ -0,0 +1,124 @@
+package memory
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// PriorityRWMutex is a mutex that allows fair access to high priority read operations while
+// blocking low priority operations if a write lock is waiting.
+type PriorityRWMutex struct {
+	lock        sync.RWMutex
+	lowPrioLock sync.RWMutex
+}
+
+type BlockContext struct {
+	lock         *PriorityRWMutex
+	preLockTime  time.Time
+	postLockTime time.Time
+	postOpTime   time.Time
+}
+
+// RLockHigh allows *guaranteed fast* (or very high priority) ops to acquire the read lock as
+// quickly as possible. Note: This can block write lock acquisition
+func (pm *PriorityRWMutex) RLockHigh() {
+	// Don't increment activeReads just grab the lock. If an active write is holding the lock it
```

Comment:
what's activeReads. i don't see a mention of this elsewhere

Reply:
Ah, that's a leftover from a previous implementation that tracked active reads. It was an interesting idea as it allowed a non-urgent write to poll for an opportunity to block fewer reads (ideally 0), but ultimately it was a bit more complex and I wasn't sure how well it would work.

Will remove